07-routing-gossip.md

```diff
@@ -1136,6 +1157,200 @@ A->D's `update_add_htlc` message would be:
+that commits to the lease parameters included in the `lease_witness`.
+
+### The `basic_funding_lease` type
+
+A `basic_funding_lease` does not provide any guarantee that the seller won't
+close the channel or increase their routing fees after the purchase, if the
+liquidity isn't actually used.
+
+The `funding_lease_fee` is paid as detailed in the `funding_lease_fee` section
+below.
+
+The seller provides an ECDSA signature in `will_fund` using the private key
+associated with their `node_id`. The data signed is:
+
+    SHA256("basic_funding_lease" || basic_funding_lease_witness)
+
+We use a tagged hash to ensure that this signature cannot be used in a
+different context.
+
+#### Requirements
+
+A node selling a `basic_funding_lease`:
+  - MUST set `min_lease_amount_sat` and `max_lease_amount_sat` to the minimum
+    and maximum amount it will contribute at this rate.
+  - MUST set `lease_fee_base_sat` to the base fee (in satoshi) it will charge.
+  - MUST set `lease_fee_basis` to the amount it will charge per contributed
+    satoshi (in basis points, ie 1/10_000).
+  - MUST set `funding_weight` to the transaction weight that will be charged.
+    It ensures that the funding node is refunded for some of the on-chain
+    fees it will pay to contribute the requested funds to a channel.
+
+### The `duration_based_funding_lease` type
+
+A `duration_based_funding_lease` commits to a duration during which the seller
+will not remove the funds nor increase their routing fees above the maximum
+values provided in the lease. It is impossible to actually prevent them from
+doing so, but the signature they provide in `will_fund` can be used to prove
+that they cheated.
+
+The `funding_lease_fee` is paid as detailed in the `funding_lease_fee` section
+below.
+
+The seller provides an ECDSA signature in `will_fund` using the private key
+associated with their `node_id`. The data signed is:
+
+    SHA256("duration_based_funding_lease" || duration_based_funding_lease_witness)
+
+We use a tagged hash to ensure that this signature cannot be used in a
+different context.
+
+#### Requirements
+
+A node selling a `duration_based_funding_lease`:
+  - MUST set `lease_duration` to the number of blocks during which the lease
+    will be active.
+  - MUST set `max_channel_fee_base_msat` to the maximum `fee_base_msat` it
+    will use in its `channel_update` while the lease is active.
+  - MUST set `max_channel_fee_basis` to match (when converted from basis
+    points to millionths) the maximum `fee_proportional_millionths` it will
+    use in its `channel_update` while the lease is active.
+
+### Paying the `funding_lease_fee`
+
+When `request_funds` and `will_fund` have been exchanged, the buyer must pay
+fees to the seller for the funding they provide to the channel based on the
+agreed upon `funding_lease_fee`.
+
+The lease fee is taken from the buyer's funding inputs and added to the
```

Comment:
In this case the client already needs funds available onchain to receive the channel. If the way the lease fee is taken would be defined by TLV types as well, there could be multiple options for payment methods. I'm thinking for example JIT channels, where the fee is taken from htlcs.

EDIT: I suppose in that case the `funding_lease_fee` would be zero and the payment method can go in a TLV elsewhere.

Reply:
In the case where the buyer doesn't have any funds (either on-chain or in an existing channel) to pay the fees, you wouldn't use one of the lease types provided here. That's where an on-the-fly liquidity bLIP would introduce a new lease type that would pay the fees by subtracting it from the HTLC amount (or from all future HTLCs), most likely based on https://github.com/lightning/blips/pull/25

Reply:
> My main point is that independently from the lease type, the client will always need to be able to calculate the withheld fee

Not necessarily: in a model where the LSP only applies a proportional fee on all HTLCs received by the wallet and automatically provides liquidity as needed, there may be no fee applied at each funding/splicing event at all. The actual fee will depend on future payments received by the user, which we don't know at funding time. I don't think we need to require that the fee has to be known beforehand, this would limit the ways this can be used?

> for any additional lease types we add we should also clearly define how the fee will be calculated for them, i.e. we'll need specific `Paying the ..` sections for each of them

Yes, I agree with that: each type of lease will need to explain the fee structure associated with it. That's why I currently created one section per lease type, in which that can be detailed. Since the two lease types I added share the same fee mechanism, I factored that in its own section, maybe that's what created some confusion? If we remove the `duration_based_funding_lease` from this version, I'll nest the `Paying the fee` section inside the `The basic_funding_lease type` section, which may be clearer?

Reply:
> Where the buyer doesn't pay any fee when the channel is created

One current limitation with channel buys using liquidity ads is that the buyer initiates the channel open, which means they're also responsible for the onchain closure fees. So the buyer *has* to contribute at least some funds to the channel at start in order for it to be operational (some CLN users have run into problems where they created basically unusable channels because they didn't put enough sats in at start for a liquidity buy)

If a buyer doesn't have any sats to begin with, they won't be able to use liqudity ad buys.

One way to solve this would be to add a flag that allows for flipping which node is considered the "opener" for the calculation of closing fees. This would put the cost/risk of channel closure onto the liquidity seller, and allows for an opener to essentially do a zero-funds (no utxo) channel open. I'd presume that the seller could price this option independently (maybe add it as a new type of lease?) as it does add a potential cost risk to them.

When you pair this with the JIT/HTLC payment plan bLIP proposals, I think you end up with a pretty nice protocol for allowing zero-UTXO parties to buy inbound liquidity at will. Maybe I'm out of date on the cost models for onchain fees in commitment transactions though?

Reply:
> One current limitation with channel buys using liquidity ads is that the buyer initiates the channel open, which means they're also responsible for the onchain closure fees. So the buyer has to contribute at least some funds to the channel at start in order for it to be operational (some CLN users have run into problems where they created basically unusable channels because they didn't put enough sats in at start for a liquidity buy)

Indeed, that's why I've been experimenting with https://github.com/ACINQ/eclair/pull/2845 as a short-term solution until we move to 0-fee commit txs thanks to v3 transactions. I planned to make that option a bLIP if we want to ship before v3 transactions.

Reply:
> Indeed, that's why I've been experimenting with https://github.com/ACINQ/eclair/pull/2845 as a short-term solution until we move to 0-fee commit txs thanks to v3 transactions. I planned to make that option a bLIP if we want to ship before v3 transactions.

Is that a good reason to not add it to the spec currently though? I presume there's a lot of other work to be done before v3 transactions are widely supported. The fact that you're experimenting with it in eclair already is a good indication that this is in fact a useful/necessary feature to add more widely?

This is a side conversation for sure but I'm still a bit unclear about the proposal for how non-utxo owning lightning nodes will manage channel closes, even in a v3 regime.

Reply:
I don't think it would belong to the BOLTs, but a bLIP is definitely needed! That's on my todo-list once I've finished the on-the-fly funding prototype, I'll open a PR for it soon. Would a bLIP work for you?